### PR TITLE
[FEAT] Drag and Drop

### PR DESCRIPTION
**Pull request by c-py**
Co-Authored-By: Spencer Dezart-Smith <17863697+spencerdezartsmith@users.noreply.github.com>

# Description

This PR implements drag and drop of buildings.
drag and drop

## Features

- Alerts when bakery is placed
- Collision detection with resources, water and map corners. (bumpkin collisions missing)
- Cancellation of placement

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

TDD and Manual testing

# Checklist:

- [X] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [X] Screenshot if it includes any UI changes
- [X] I have read the contributing guidelines and agree to the T&Cs
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/features/farming/hud/Hud.tsx [hunk 1]
```diff
@@ -29,6 +30,7 @@ export const Hud: React.FC = () => {
       />
       <AudioPlayer isFarming />
       <VisitBanner id={landId} />
+      <Buildings />
```
Comment (c-py) [inline src/features/farming/hud/Hud.tsx:33]: This is a temporary addition to the HUD to allow building placements.

File: src/features/game/expansion/Land.tsx [hunk 2]
```diff
@@ -118,17 +120,31 @@ const Expansion: React.FC<ExpansionProps & { expansionIndex: number }> = ({
...
+
+          return (
+            <MapPlacement
+              key={`${createdAt}-stone-${index}`}
+              x={x + xOffset}
+              y={y + yOffset}
+              height={height}
+              width={width}
+            >
+              <div style={{ background: "black" }} className="w-full h-full" />
```
Comment (c-py) [inline src/features/game/expansion/Land.tsx:136]: A stone is currently represented by a black div as stones have not been implemented yet.

File: src/features/game/expansion/Land.tsx [hunk 3]
```diff
@@ -147,20 +163,37 @@ export const Land: React.FC = () => {
...
+            className="absolute"
+            style={{
+              top: "50%",
+              left: "50%",
+              zIndex: 100,
+            }}
+          >
+            <Placeable />
+          </div>
+        )}
```
Comment (c-py) [inline src/features/game/expansion/Land.tsx:195]: A placeable can be a building, decoration or NFT. They are hardcoded here to start drag and drop in them middle of the screen.

File: src/features/game/expansion/LandExpansion.tsx [hunk 4]
```diff
@@ -45,6 +45,7 @@ export const LandExpansion: React.FC = () => {
         <ScrollContainer
           className="bg-blue-300 overflow-scroll relative w-full h-full"
           innerRef={container}
+          ignoreElements={"*[data-prevent-drag-scroll]"}
```
Comment (c-py) [inline src/features/game/expansion/LandExpansion.tsx:48]: This allows us to set the data attribute `data-prevent-drag-scroll` on any element. This will prevent dragging the map scroll container when hovering on the element.

File: src/features/game/expansion/components/Water.tsx [hunk 5]
```diff
@@ -31,10 +30,10 @@ export const Water: React.FC<Props> = ({ level }) => {
       className="absolute inset-0"
     >
       {/* Above Land */}
-      <Shark side="top" />
+      {/* <Shark side="top" /> */}
 
       {/* Below Land */}
-      <Shark side="bottom" />
+      {/* <Shark side="bottom" /> */}
```
Comment (c-py) [inline src/features/game/expansion/components/Water.tsx:36]: The shark is dead to us.

File: src/features/game/expansion/placeable/lib/collisionDetection.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+ * Extracts the positional data from all the resources on all land expansions.
+ * @param expansions
+ * @returns Array of all resource positions
+ */
+export function extractResourcePositions(
+  expansions: Required<
+    Omit<LandExpansion, "terrains" | "createdAt" | "readyAt">
+  >[]
+) {
+  return expansions.flatMap((expansion, expansionIndex) =>
```
Comment (adamhannigan) [inline src/features/game/expansion/placeable/lib/collisionDetection.ts:36]: Can we break this apart. It is a bit hard to follow, I'm not sure what is going on here 😅 

`flatMap` -> `Object.values` -> `omit` -> `flatMap` -> `extract`

File: src/features/game/expansion/placeable/lib/collisionDetection.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+  const expansionPositions: Position[] = expansions.map(
+    (_, expansionIndex) => ({
+      x: EXPANSION_ORIGINS[expansionIndex].x - LAND_SIZE / 2,
+      y: EXPANSION_ORIGINS[expansionIndex].y + LAND_SIZE / 2,
+      width: LAND_SIZE,
+      height: LAND_SIZE,
+    })
+  );
+
+  // Split a MxN position into 1x1 chunks
```
Comment (adamhannigan) [inline src/features/game/expansion/placeable/lib/collisionDetection.ts:93]: Why are we splitting Mexican Pesos? What does MxN mean 😮

File: src/features/game/expansion/placeable/lib/collisionDetection.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+function detectResourceCollision(state: GameState, position: Position) {
+  const { expansions } = state;
+
+  const resourcesFromExpansions = expansions.map((expansion) => ({
+    shrubs: expansion.shrubs ?? {},
+    pebbles: expansion.pebbles ?? {},
+    trees: expansion.trees ?? {},
+    stones: expansion.stones ?? {},
+    plots: expansion.plots ?? {},
+  }));
```
Comment (adamhannigan) [inline src/features/game/expansion/placeable/lib/collisionDetection.ts:72]: Have you thought about detecting other buildings that have been placed?

File: src/features/game/expansion/placeable/lib/collisionDetection.ts (hunk 7 above)
Comment (adamhannigan) [reply src/features/game/expansion/placeable/lib/collisionDetection.ts:93]: Consider "Transform a 2D array into a 1D array".

Chunks is a bit generic, could we name it something more detailed - is this representing a Grid Square?

**Review by adamhannigan — COMMENTED** [review]
Nice work guys!

We were having a bit of trouble understanding some of the collision detection. If possible could we add some more comments and detailed variable names to help break down what is happening.

Also, did you get a chance to do any performance testing? How do you feel these functions would go with 10 land expansions, 50 resources and 15 buildings?

File: src/features/game/expansion/placeable/lib/collisionDetection.ts (hunk 8 above)
Comment (spencerdezartsmith) [reply src/features/game/expansion/placeable/lib/collisionDetection.ts:97]: We will implement this once we understand the data structure of buildings.

**Comment by spencerdezartsmith** [comment]
> Nice work guys!
> 
> We were having a bit of trouble understanding some of the collision detection. If possible could we add some more comments and detailed variable names to help break down what is happening.
> 
> Also, did you get a chance to do any performance testing? How do you feel these functions would go with 10 land expansions, 50 resources and 15 buildings?

@c-py said performance is fine 😉

File: src/features/game/expansion/placeable/lib/collisionDetection.ts [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+      case Direction.Right:
+        return expansionExistsAtOffset(origin, { x: 1, y: 0 });
+      case Direction.Top:
+        return expansionExistsAtOffset(origin, { x: 0, y: 1 });
+      case Direction.Bottom:
+        return expansionExistsAtOffset(origin, { x: 0, y: -1 });
+    }
+  };
+
+  return originCoordiatesForExpansions.some((originCoordinate) => {
```
Comment (Rofrtd) [inline src/features/game/expansion/placeable/lib/collisionDetection.ts:202]: `originCoordiatesForExpansions` typo on coordinates.

**Review by Rofrtd — APPROVED** [review]
LGTM. 
Awesome work fellas.
Just a typo to be fixed.
🚀
